scripts/tiling/stitching.py

```diff
@@ -0,0 +1,256 @@
+
+
+def run_task(fov_paths, out_path, session_dict, mt_upload):
+    unique_rois = np.unique([os.path.basename(p).split("-")[-1].split(".tiff")[0].split("_")[0] for p in fov_paths])
+    roi_path_groups = dict([(u,[p for p in fov_paths if u in p]) for u in unique_rois])
+    mr = None
+    for t in range(MAX_TRIES):
+        try:
+            mr = MibiRequests(**session_dict)
+            break
+        except:
+            if t < MAX_TRIES-1:
+                time.sleep(0.50)
+            else:
+                mr = MibiRequests(**session_dict)
+
+    for roi,roi_fov_paths in roi_path_groups.items():
+        print(roi)
+
+        cols = []
+        rows = []
+        for fov_path in roi_fov_paths:
+            fov_name = os.path.basename(fov_path).split("-")[-1].split(".tiff")[0]
+            c, r = fov_name.split("_")[1:]
+            cols.append(int(c[1:].lstrip("0")))
+            rows.append(int(r[1:].lstrip("0")))
+
+        out_img, panel, max_dim = combine_entity_by_name(roi_fov_paths, cols, rows, enforce_square=True)
+        out_img = out_img.astype(np.uint8, copy=False)
+
+        um_min_x, um_min_y = 999999999, 999999999
+        for fov_path in roi_fov_paths:
+            
+            json_file = os.path.dirname(fov_path)+"/"+"-".join(os.path.basename(fov_path).split("-")[:2])+"-scan-1.json"
+            with open(json_file) as f:
+                bin_json = json.load(f)
+                coord = bin_json["coordinates"]
+                if coord["x"] < um_min_x:
+                    um_min_x = coord["x"]
+                if coord["y"] < um_min_y:
+                    um_min_y = coord["y"]
+
+        run = os.path.basename(os.path.dirname(roi_fov_paths[0]))
+        fov = "FOV"+os.path.basename(roi_fov_paths[0]).split("-")[1]
+        print(run, fov)
+        
+        for t in range(MAX_TRIES):
+            try:
+                try:
+                    ref_image_id = mr.image_id(run, fov)
+                except:
+                    ref_image_id = mr.image_id(run, fov.replace("V0", "V"))
+                break
+            except:
+                if t < MAX_TRIES-1:
+                    time.sleep(0.50)
+                else:
+                    try:
+                        ref_image_id = mr.image_id(run, fov)
+                    except:
+                        ref_image_id = mr.image_id(run, fov.replace("V0", "V"))
+        
+        for t in range(MAX_TRIES):
+            try:
+                ref_json = mr.get('images/{}/'.format(ref_image_id)).json()
+                break
+            except:
+                if t < MAX_TRIES-1:
+                    time.sleep(0.50)
+                else:
+                    ref_json = mr.get('images/{}/'.format(ref_image_id)).json()
+
+        sample_id = run+"__"+roi
+        f_split = ref_json['folder'].split('/')
+        f_split[0] = sample_id
+        folder_name = '/'.join(f_split[:3])
+
+        px_per_u = ref_json["frame"]/ref_json["fov_size"]
+
+        metadata = {
```

Comment:
Is this really needed? The majority of metadata should be the same across all FOVs within a ROI (except coordinates, obviously). And in that case, you could just copy the metadata from the first TIFF file you read in.

Reply:
Hmm, yeah, good point. That will mean we won't need to connect to MIBItracker at all if we are not uploading to it.